File: Robotics_Kinematics_and_Dynamics_Course/Chapter7/Lesson3/Chapter7_Lesson3.cpp
#include <iostream>
#include <vector>
#include <Eigen/Dense>

using Eigen::Matrix3d;
using Eigen::Matrix4d;
using Eigen::MatrixXd;
using Eigen::Vector3d;
using Eigen::VectorXd;

Matrix3d skew(const Vector3d& v) {
    Matrix3d S;
    S <<     0.0, -v(2),  v(1),
           v(2),    0.0, -v(0),
          -v(1),  v(0),   0.0;
    return S;
}

Matrix4d se3Hat(const VectorXd& S) {
    Matrix4d M = Matrix4d::Zero();
    Vector3d omega = S.segment<3>(0);
    Vector3d v = S.segment<3>(3);
    M.block<3,3>(0,0) = skew(omega);
    M.block<3,1>(0,3) = v;
    return M;
}

// Exponential for twist S and scalar theta
Matrix4d se3Exp(const VectorXd& S, double theta, double tol=1e-9) {
    Matrix4d result = Matrix4d::Identity();
    Vector3d omega = S.segment<3>(0);
    Vector3d v = S.segment<3>(3);
    double wnorm = omega.norm();
    if (wnorm < tol) {
        // Pure translation
        result.block<3,1>(0,3) = v * theta;
        return result;
    }
    Vector3d w = omega / wnorm;
    Matrix3d w_hat = skew(w);
    double th = wnorm * theta;
    Matrix3d R = Matrix3d::Identity()
                 + std::sin(th) * w_hat
                 + (1.0 - std::cos(th)) * (w_hat * w_hat);
    Matrix3d G = Matrix3d::Identity() * th
                 + (1.0 - std::cos(th)) * w_hat
                 + (th - std::sin(th)) * (w_hat * w_hat);
    Vector3d p = G * (v / wnorm);
    result.block<3,3>(0,0) = R;
    result.block<3,1>(0,3) = p;
    return result;
}

Eigen::Matrix<double,6,6> adjoint(const Matrix4d& T) {
    Eigen::Matrix<double,6,6> Ad = Eigen::Matrix<double,6,6>::Zero();
    Matrix3d R = T.block<3,3>(0,0);
    Vector3d p = T.block<3,1>(0,3);
    Ad.block<3,3>(0,0) = R;
    Ad.block<3,3>(3,3) = R;
    Ad.block<3,3>(3,0) = skew(p) * R;
    return Ad;
}

Matrix4d fkPoeSpace(const std::vector<VectorXd>& Slist,
                    const Matrix4d& M,
                    const VectorXd& q) {
    Matrix4d T = Matrix4d::Identity();
    for (int i = 0; i < (int)Slist.size(); ++i) {
        T = T * se3Exp(Slist[i], q(i));
    }
    T = T * M;
    return T;
}

MatrixXd jacobianSpace(const std::vector<VectorXd>& Slist,
                       const VectorXd& q) {
    int n = (int)Slist.size();
    MatrixXd J(6, n);
    J.setZero();
    Matrix4d T = Matrix4d::Identity();
    for (int i = 0; i < n; ++i) {
        if (i == 0) {
            J.col(0) = Slist[0];
        } else {
            Eigen::Matrix<double,6,6> AdT = adjoint(T);
            J.col(i) = AdT * Slist[i];
        }
        T = T * se3Exp(Slist[i], q(i));
    }
    return J;
}

MatrixXd jacobianBody(const std::vector<VectorXd>& Slist,
                      const Matrix4d& M,
                      const VectorXd& q) {
    Matrix4d T = fkPoeSpace(Slist, M, q);
    MatrixXd Js = jacobianSpace(Slist, q);
    Eigen::Matrix<double,6,6> AdInv = adjoint(T.inverse());
    return AdInv * Js;
}

int main() {
    double l1 = 1.0, l2 = 0.8;
    VectorXd S1(6), S2(6);
    S1 << 0, 0, 1, 0, 0, 0;
    S2 << 0, 0, 1, 0, -l1, 0;
    std::vector<VectorXd> Slist = {S1, S2};

    Matrix4d M = Matrix4d::Identity();
    M(0,3) = l1 + l2;

    VectorXd q(2);
    q << 0.5, -0.4;

    Matrix4d T = fkPoeSpace(Slist, M, q);
    MatrixXd Js = jacobianSpace(Slist, q);
    MatrixXd Jb = jacobianBody(Slist, M, q);

    std::cout << "T(q):\n" << T << std::endl;
    std::cout << "J_s(q):\n" << Js << std::endl;
    std::cout << "J_b(q):\n" << Jb << std::endl;
    return 0;
}
      

File: Robotics_Kinematics_and_Dynamics_Course/Chapter7/Lesson3/Chapter7_Lesson3.java
import org.ejml.simple.SimpleMatrix;

public class JacobianSE3 {

    public static SimpleMatrix skew(SimpleMatrix v) {
        double vx = v.get(0), vy = v.get(1), vz = v.get(2);
        double[][] data = {
            {0.0, -vz,  vy},
            {vz,  0.0, -vx},
            {-vy, vx,   0.0}
        };
        return new SimpleMatrix(data);
    }

    public static SimpleMatrix se3Hat(SimpleMatrix S) {
        SimpleMatrix M = SimpleMatrix.identity(4);
        M.zero();
        SimpleMatrix omega = S.rows(0, 3);
        SimpleMatrix v = S.rows(3, 6);
        M.insertIntoThis(0, 0, skew(omega));
        M.insertIntoThis(0, 3, v);
        return M;
    }

    public static SimpleMatrix se3Exp(SimpleMatrix S, double theta) {
        double tol = 1e-9;
        SimpleMatrix omega = S.rows(0, 3);
        SimpleMatrix v = S.rows(3, 6);
        double wnorm = omega.normF();
        SimpleMatrix T = SimpleMatrix.identity(4);
        if (wnorm < tol) {
            // pure translation
            SimpleMatrix p = v.scale(theta);
            T.insertIntoThis(0, 3, p);
            return T;
        }
        SimpleMatrix w = omega.scale(1.0 / wnorm);
        SimpleMatrix w_hat = skew(w);
        double th = wnorm * theta;

        SimpleMatrix I3 = SimpleMatrix.identity(3);
        SimpleMatrix R = I3.plus(w_hat.scale(Math.sin(th)))
                           .plus(w_hat.mult(w_hat).scale(1.0 - Math.cos(th)));

        SimpleMatrix G = I3.scale(th)
                .plus(w_hat.scale(1.0 - Math.cos(th)))
                .plus(w_hat.mult(w_hat).scale(th - Math.sin(th)));

        SimpleMatrix p = G.mult(v.scale(1.0 / wnorm));
        T.insertIntoThis(0, 0, R);
        T.insertIntoThis(0, 3, p);
        return T;
    }

    public static SimpleMatrix adjoint(SimpleMatrix T) {
        SimpleMatrix R = T.extractMatrix(0, 3, 0, 3);
        SimpleMatrix p = T.extractMatrix(0, 3, 3, 4);
        SimpleMatrix Ad = new SimpleMatrix(6, 6);
        Ad.insertIntoThis(0, 0, R);
        Ad.insertIntoThis(3, 3, R);
        Ad.insertIntoThis(3, 0, skew(p).mult(R));
        return Ad;
    }

    public static SimpleMatrix fkPoeSpace(SimpleMatrix[] Slist,
                                          SimpleMatrix M,
                                          double[] q) {
        SimpleMatrix T = SimpleMatrix.identity(4);
        for (int i = 0; i < Slist.length; ++i) {
            T = T.mult(se3Exp(Slist[i], q[i]));
        }
        return T.mult(M);
    }

    public static SimpleMatrix jacobianSpace(SimpleMatrix[] Slist, double[] q) {
        int n = Slist.length;
        SimpleMatrix J = new SimpleMatrix(6, n);
        SimpleMatrix T = SimpleMatrix.identity(4);
        for (int i = 0; i < n; ++i) {
            if (i == 0) {
                J.insertIntoThis(0, 0, Slist[0]);
            } else {
                SimpleMatrix AdT = adjoint(T);
                SimpleMatrix col = AdT.mult(Slist[i]);
                J.insertIntoThis(0, i, col);
            }
            T = T.mult(se3Exp(Slist[i], q[i]));
        }
        return J;
    }

    public static SimpleMatrix jacobianBody(SimpleMatrix[] Slist,
                                            SimpleMatrix M,
                                            double[] q) {
        SimpleMatrix T = fkPoeSpace(Slist, M, q);
        SimpleMatrix Js = jacobianSpace(Slist, q);
        SimpleMatrix AdInv = adjoint(T.invert());
        return AdInv.mult(Js);
    }

    public static void main(String[] args) {
        double l1 = 1.0, l2 = 0.8;
        SimpleMatrix S1 = new SimpleMatrix(6, 1, true,
                new double[]{0, 0, 1, 0, 0, 0});
        SimpleMatrix S2 = new SimpleMatrix(6, 1, true,
                new double[]{0, 0, 1, 0, -l1, 0});
        SimpleMatrix[] Slist = new SimpleMatrix[]{S1, S2};

        SimpleMatrix M = SimpleMatrix.identity(4);
        M.set(0, 3, l1 + l2);

        double[] q = new double[]{0.5, -0.4};
        SimpleMatrix T = fkPoeSpace(Slist, M, q);
        SimpleMatrix Js = jacobianSpace(Slist, q);
        SimpleMatrix Jb = jacobianBody(Slist, M, q);

        System.out.println("T(q) = \n" + T);
        System.out.println("J_s(q) = \n" + Js);
        System.out.println("J_b(q) = \n" + Jb);
    }
}
      
